cmd/snap-seccomp/main_test.go

```diff
@@ -0,0 +1,404 @@
+		for i := range args {
+			if nr, err := strconv.ParseUint(args[i], 10, 64); err == nil {
+				syscallArgs[i] = nr
+			} else {
+				syscallArgs[i] = main.SeccompResolver[args[i]]
+			}
+		}
+	}
+	sd := &main.SeccompData{}
+	sd.SetArch(arch)
+	sd.SetNr(sc)
+	sd.SetArgs(syscallArgs)
+	return sd, nil
+}
+
+// Endianness detection.
+func nativeEndian() binary.ByteOrder {
+	// Credit matt kane, taken from his gosndfile project.
+	// https://groups.google.com/forum/#!msg/golang-nuts/3GEzwKfRRQw/D1bMbFP-ClAJ
+	// https://github.com/mkb218/gosndfile
+	var i int32 = 0x01020304
+	u := unsafe.Pointer(&i)
+	pb := (*byte)(u)
+	b := *pb
+	isLittleEndian := b == 0x04
+
+	if isLittleEndian {
+		return binary.LittleEndian
+	} else {
+		return binary.BigEndian
+	}
+}
+
+func simulateBpf(c *C, seccompWhitelist, bpfInput string, expected int) {
+	outPath := filepath.Join(c.MkDir(), "bpf")
+	err := main.Compile([]byte(seccompWhitelist), outPath)
+	c.Assert(err, IsNil)
+
+	ops, err := decodeBpfFromFile(outPath)
+	c.Assert(err, IsNil)
+
+	vm, err := bpf.NewVM(ops)
+	c.Assert(err, IsNil)
+
+	bpfSeccompInput, err := parseBpfInput(bpfInput)
+	c.Assert(err, IsNil)
+
+	buf2 := (*[64]byte)(unsafe.Pointer(bpfSeccompInput))
+
+	out, err := vm.Run(buf2[:])
+	c.Assert(err, IsNil)
+	c.Check(out, Equals, expected, Commentf("unexpected result for %q (input %q), got %v expected %v", seccompWhitelist, bpfInput, out, expected))
+}
+
+func (s *snapSeccompSuite) TestCompile(c *C) {
+	// FIXME: this will only work once the following issue is fixed:
+	// https://github.com/golang/go/issues/20556
+	bpf.VmEndianness = nativeEndian()
+
+	for _, t := range []struct {
+		seccompWhitelist string
+		bpfInput         string
+		expected         int
+	}{
+		// special
+		{"@unrestricted", "execve", main.SeccompRetAllow},
+		{"@complain", "execve", main.SeccompRetAllow},
+
+		// trivial alllow
+		{"read", "read", main.SeccompRetAllow},
+		{"read\nwrite\nexecve\n", "write", main.SeccompRetAllow},
+
+		// trivial denial
+		{"read", "execve", main.SeccompRetKill},
+
+		// argument filtering
+		{"read >=2", "read;native;2", main.SeccompRetAllow},
+		{"read >=2", "read;native;1", main.SeccompRetKill},
+
+		{"read <=2", "read;native;2", main.SeccompRetAllow},
+		{"read <=2", "read;native;3", main.SeccompRetKill},
+
+		{"read !2", "read;native;3", main.SeccompRetAllow},
+		{"read !2", "read;native;2", main.SeccompRetKill},
+
+		{"read >2", "read;native;2", main.SeccompRetKill},
+		{"read >2", "read;native;3", main.SeccompRetAllow},
+
+		{"read <2", "read;native;2", main.SeccompRetKill},
+		{"read <2", "read;native;1", main.SeccompRetAllow},
+
+		// FIXME: test maskedEqual better
+		{"read |1", "read;native;1", main.SeccompRetAllow},
+		{"read |1", "read;native;2", main.SeccompRetKill},
+
+		{"read 2", "read;native;3", main.SeccompRetKill},
+		{"read 2", "read;native;2", main.SeccompRetAllow},
+
+		// with arg1 and name resolving
+		{"ioctl - TIOCSTI", "ioctl;native;0,TIOCSTI", main.SeccompRetAllow},
```

Comment:
With `ioctl - TIOCSTI` the first argument is ignored in the generated bpf. The `ioctl;native;0,TIOCSTI` input is just one of the possible inputs that are legal. I could replace the `0` in there with a `1` or a `42` and it would still pass. I'm happy to add support for `-` in the input as well. The `-` would mean a random in then and we would have better tests because of this.